src/actions/editor.ts

```diff
@@ -0,0 +1,288 @@
+import { ActionBase, AppThunk } from "../lib/store";
+import { showNotification } from "./notifications";
+import { NotificationLevel } from "../models/notification";
+import { isValidConnection } from "../lib/editorUtils";
+import { getConnection, getConnectionByNodesAndPorts, getConnectionsForSinkNodeAndPort, getConnectionsForSourceNodeAndPort, getNode } from "../selectors/graph";
+import { GraphConnectionRecord, NodeType } from "../models/graph";
+import { EditorNodeRecord } from "../models/editor";
+import { setInstanceSourcePortConnections, setInstanceSinkPortConnections, unloadPatcherFromRemoteInstance } from "./device";
+
+export enum EditorActionType {
+	POSITION_NODE = "POSITION_NODE",
+	SELECT_NODE = "SELECT_NODE",
+	UNSELECT_NODE = "UNSELECT_NODE",
+	SELECT_EDGE = "SELECT_EDGE",
+	UNSELECT_EDGE = "UNSELECT_EDGE"
+}
+
+export interface IPositionNode extends ActionBase {
+	type: EditorActionType.POSITION_NODE;
+	payload: {
+		id: EditorNodeRecord["id"];
+		x: number;
+		y: number;
+	};
+}
+
+export interface ISelectNode extends ActionBase {
+	type: EditorActionType.SELECT_NODE;
+	payload: {
+		id: EditorNodeRecord["id"];
+	};
+}
+
+export interface IUnselectNode extends ActionBase {
+	type: EditorActionType.UNSELECT_NODE;
+	payload: {
+		id: EditorNodeRecord["id"];
+	};
+}
+
+export interface ISelectEgde extends ActionBase {
+	type: EditorActionType.SELECT_EDGE;
+	payload: {
+		id: EditorNodeRecord["id"];
+	};
+}
+
+export interface IUnselectEdge extends ActionBase {
+	type: EditorActionType.UNSELECT_EDGE;
+	payload: {
+		id: EditorNodeRecord["id"];
+	};
+}
+
+export type EditorAction = IPositionNode | ISelectNode | IUnselectNode | ISelectEgde | IUnselectEdge;
+
+export const makeEditorConnection = (connection: Connection): AppThunk =>
+	(dispatch, getState) => {
+		try {
+			const state = getState();
+
+			if (!connection.source || !connection.target || !connection.sourceHandle || !connection.targetHandle) {
+				throw new Error(`Invalid Connection Description (${connection.source}:${connection.sourceHandle} => ${connection.target}:${connection.targetHandle})`);
+			}
+
+			// Valid Connection?
+			const { sourceNode, sourcePort, sinkNode, sinkPort } = isValidConnection(connection, state.graph.nodes);
+
+			// Does it already exist?
+			const existingConnection = getConnectionByNodesAndPorts(
+				state,
+				{
+					sourceNodeId: sourceNode.id,
+					sinkNodeId: sinkNode.id,
+					sourcePortId: sourcePort.id,
+					sinkPortId: sinkPort.id
+				}
+			);
+
+			if (existingConnection) {
+				return void dispatch(showNotification({
+					title: "Skipped creating connection",
+					level: NotificationLevel.warn,
+					message: `A connection between ${connection.source}:${connection.sourceHandle} and ${connection.target}:${connection.targetHandle} already exists`
+				}));
+			}
+
+			if (sourceNode.type === NodeType.System && sinkNode.type === NodeType.System) {
+				// TODO: Discuss whether we allow passthrough
```

Comment:
yeah.. and, we might eventually have non-rnbo non-system nodes that we could represent, other applications, you could hook to/from

Reply:
Yeah so there is no passthrough implemented as the information about system connections isn't clear to me where / how it described. the `/rnbo/jack` JSON doesn't really describe the connections (in contrast to the instances) so I simply didn't implement it thus far.